File: design_pattern/builder_design_pattern/student.cpp
#include<iostream>
#include<string>
#include<vector>
using namespace std;


class Student;  // Forward declaration
// create a abstract student builder class
class StudentBuilder {
    public:
        string name, email;
        int rollNo,age;
        vector<string>subjects;
        StudentBuilder* setName(string name) {
            this->name = name;
            return this;
        }
        StudentBuilder* setEmail(string email) {
            this->email = email;
            return this;
        }
        StudentBuilder* setRollNo(int rollNo) {
            this->rollNo = rollNo;
            return this;
        }
        StudentBuilder* setAge(int age) {
            this->age = age;
            return this;
        }
        StudentBuilder* addSubject(string sub){
            this->subjects.push_back(sub);
            return this;
        }
        Student* build();

        ~StudentBuilder() {
            // Destructor to clean up resources if needed
        }
};

class Student{
    string name, email;
    int rollNo,age;
    vector<string>subjects;
    public:
        Student(StudentBuilder* builder) {
            this->name = builder->name;
            this->email = builder->email;
            this->rollNo = builder->rollNo;
            this->age = builder->age;
            this->subjects = builder->subjects;
        }
        void display() {
            cout << "Name: " << name << endl;
            cout << "Email: " << email << endl;
            cout << "Roll No: " << rollNo << endl;
            cout << "Age: " << age << endl;
            cout << "Subjects: ";
            for (const auto& subject : subjects) {
                cout << subject << " ";
            }
            cout << endl;
        }
};

Student* StudentBuilder::build() {
    return new Student(this);
}

int main(){
    Student* student = (new StudentBuilder())
    ->setName("John Doe")
    ->setRollNo(12345)
    ->setAge(20)
    ->addSubject("Mathematics")
    ->build();
    student->display();
    delete student; // Clean up memory

}

File: design_pattern/builder_design_pattern/student.java
package builder_design_pattern;

public class User {
    private final String name;      // required
    private final int age;          // optional
    private final String email;     // optional
    private final String address;   // optional

    private User(Builder builder) {
        this.name = builder.name;
        this.age = builder.age;
        this.email = builder.email;
        this.address = builder.address;
    }

    public static class Builder {
        private final String name;  // required
        private int age;            // optional default 0
        private String email;
        private String address;

        public Builder(String name) {
            this.name = name;
        }

        public Builder age(int age) {
            this.age = age;
            return this;
        }

        public Builder email(String email) {
            this.email = email;
            return this;
        }

        public Builder address(String address) {
            this.address = address;
            return this;
        }

        public User build() {
            if (name == null || name.isBlank()) {
                throw new IllegalArgumentException("name is required");
            }
            return new User(this);
        }
    }
}

public class student {
    public static void main(String[] args) {
        User u = new User.Builder("Princsin")
        .age(25)
        .email("p@example.com")
        .build();
    }
}
